### ISSUE TITLE
:zap: Improve the performance of the `tui` app

### ISSUE THREAD

**Issue by neon-mmd** [maxcurzi/tplay#48]
#### Work Expected From The Issue
Provide several optimizations to improve the performance of the video playback by making the following changes:
- [ ] Use an alternative allocator like `mimalloc` to improve the performance of the TUI app.
- [ ] Enable gzip, brotli and http dynamic window size features on the reqwest client to speed up downloads.
- [ ] Use rayon to parallalize coloring and drawing work of frames both of videos and images to speed up video playback performance dramatically. 

The issue expects the following files to be changed/modified:
- `src/common/errors.rs`
- `src/main.rs`
- `src/pipeline/frames.rs`
- `src/pipeline/image_pipeline.rs`
- `src/pipeline/runner.rs`
- `src/terminal/mod.rs`
- `Cargo.toml`

> [!Note]
> All the files that are expected to be changed are located under the codebase (`tplay` directory).

#### Reason Behind These Changes
The reason behind having these changes is to improve the performance of the TUI app which can improve the performance of downloads and video playback, which can drastically improve the user experience.

#### Do you want to work on this issue?

Yes

**Comment by neon-mmd** [comment]
@maxcurzi could you please assign this issue to me. Would love to work on this issue. :slightly_smiling_face:

**Comment by maxcurzi** [comment]
@neon-mmd happy for you to have a go at it! 

Please keep things as simple as possible and measure performance before/after each update 😉

**Comment by maxcurzi** [comment]
@neon-mmd when you say "TUI app", do you mean there's work done to make `tplay` an app that can run within https://github.com/ratatui/ratatui ?
It would be cool

**Comment by neon-mmd** [comment]
> @neon-mmd I'm happy for you to have a go at it!
> 
> Please keep things as simple as possible and measure performance before/after each update 😉

Yes sure, 100% I think it is a must like without that you might optimize code without gaining any benefits. :slightly_smiling_face:

**Comment by neon-mmd** [comment]
> @neon-mmd when you say "TUI app", do you mean there's work done to make `tplay` an app that can run within https://github.com/ratatui/ratatui ? It would be cool

No, actually I didn't knew what should I call it in the issue. So that's why I called it a `tui` app but I think I can work on it like it would be cool thing to implement. :slightly_smiling_face: 

(Also, I have a few issues in mind which I will be opening soon like I found a replacement for `opencv` which I think could be great if we switch to it. As well as I am thinking to implement video preload cache like how we see on mpv or vlc where it loads and process the video ahead of time like I have an implementation in mind which will use queues. The only problem I face with video precache is that I will need to figure out how to implement a 5 seconds wait for the initiial cache load. Additionally, I will also open a feature request about looping support for the app. :slightly_smiling_face: )
